yt/frontends/art/io.py

```diff
@@ -85,7 +85,7 @@ def _read_particle_coords(self, chunks, ptf):
                 x = self._get_field((ptype, "particle_position_x"))
                 y = self._get_field((ptype, "particle_position_y"))
                 z = self._get_field((ptype, "particle_position_z"))
-                yield ptype, (x, y, z)
+                yield ptype, (x, y, z), 0.0
```

Comment:
Looks like you haven't change the Base class here, but still the behaviour changes ?
Does this mean that _every_ third party frontend that implements `_read_particle_fields` is broken by this change ?

Reply:
Yes, *however*, I did a quick check and we have extremely few if *any* that implement this, unless they are hidden somewhere.  Because this was introduced in yt-4.0, it should also be mitigated.  Would a wrapper around the functions via metaclass work for you?

Reply:
Since we haven't been promoting (yet) the recipe for third party frontends, it's indeed unlikely that this actually breaks anyone, so I don't think we need to change the way you're doing it, and a quick public announcement (an email to yt-dev) should be more than enough.

Reply:
I was in a seminar today where @gvwilson talked about the "contracts" we have with software from a design perspective, and how we should only ever *relax* our specifications on what is input to a function and *strengthen* what is allowed to be output by a function.  This violates that principle, and I am thinking about it in a new way.  So let me think a bit.

Reply:
@matthewturk any further thoughts on this? 

I think if we didn't want to add the 0 `hsml` to all the `_read_particle_coords` methods, we could instead check the length of the returned tuple where needed? For example in `BaseParticleIOHanlder._count_particles_chunks` we have:

```python
            for ptype, (x, y, z), hsml in self._read_particle_coords(chunks, ptf):
                psize[ptype] += selector.count_points(x, y, z, hsml) 
```
but could do something along the lines of 

```python
            for coord_tuple in self._read_particle_coords(chunks, ptf):
                ptype = coord_tuple[0]
                x, y, z = coord_tuple[1]
                if len(coord_tuple) == 3:
                    hsml = coord_tuple[2]
                else:
                    hsml = 0.
                psize[ptype] += selector.count_points(x, y, z, hsml)
```

Reply:
Note: we decided to keep the hsml in all the `_read_particle_coords` (the base class now reflects this change). Having the different behavior with variable outputs proved difficult with the type checking and overall more complicated, so I think making this change while there are few if any third party plugins to break is the way to go.